### PR TITLE
CAS-186/update referral dates

### PR DESCRIPTION
**Pull request by froddd**
# Context

https://dsdmoj.atlassian.net/browse/CAS-186

# Changes in this PR

Introduces the ability for an HPT to change the Release date and Accommodation required from date for a given referral:

- Adds a 'Change' link next to both Release date and Accommodation required from date on the HPT full referral summary
- When clicked, opens a new screen to change the selected date
- Shows a success banner on referral summary when date has been updated
- The updated 'accommodation required' date is used everywhere it is shown:
  - Full referral summary, shown as 'Accommodation required from date'
  - 'Place context' component when searching for a bedspace, shown as 'Accommodation required from: {date}'
  - List of referrals (Unallocated/In review/Ready to place/Archived), shown under 'Bedspace required' column

### Full referral view

Screenshot 2024-07-09 at 10 19 20

### 'Accommodation required from date' edit form, showing error

Screenshot 2024-07-09 at 10 19 47

### Success banner on referral summary page

Screenshot 2024-07-09 at 10 20 10

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work? Yes, https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/2035
- [ ] Have they been released to production already? Nope

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

### CODE REVIEW

File: server/views/temporary-accommodation/assessments/change-date.njk [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    {% set hintHtml %}
+        {% if content.hint %}
+            <p class="govuk-hint">{{ content.hint }}</p>
+        {% endif %}
+        <p class="govuk-hint">{{ dateInputHint('future') }}</p>
+    {% endset %}
+
+    <h1 class="govuk-heading-l">{{ content.title }}</h1>
+
+    {{ showErrorSummary(errorSummary) }}
```
Comment (aliuk2012) [inline server/views/temporary-accommodation/assessments/change-date.njk]: If I understand the code correctly, shouldn't the error summary be above the h1?

https://design-system.service.gov.uk/components/error-summary/#where-to-put-the-error-summary

Comment (froddd) [reply server/views/temporary-accommodation/assessments/change-date.njk]: I believe you are correct, and for some reason this pattern was not adopted for this service (a quick search shows all error summaries show _after_ the h1) -- probably one to fix as a separate ticket!

**Comment by aliuk2012** [comment]
Minor issue that might be worth bringing up with content/ux designers

image
according to https://design-system.service.gov.uk/components/error-summary/#how-it-works

>- link to each of the answers that have validation errors
>- make sure the error messages in the error summary are worded the same as those which appear next to the inputs with errors

**Comment by froddd** [comment]
These custom error messages are as designed, but it’s a good point — there’s definitely room for improvement here. I’ll have a chat with the designers to see how they can be made better.

**Review by aliuk2012 — APPROVED** [review]
Works locally for me 👍

File: server/utils/applicationUtils.ts [hunk 2]
```diff
@@ -176,12 +185,11 @@ const taskResponsesToSummaryListRowItems = (
                 ? formatLines(value as string)
                 : embeddedSummaryListItem(value as Array<Record<string, unknown>>),
           },
+          actions: actionItems ? { items: actionItems } : undefined,
```
Comment (richpjames) [inline server/utils/applicationUtils.ts:190]: should this be checking for `actionItems.length`?

**Review by richpjames — APPROVED** [review]

File: server/utils/applicationUtils.ts (hunk 2 above)
Comment (froddd) [reply server/utils/applicationUtils.ts:190]: Nope, `actionItems` is either an array of `SummaryListActionItem` or undefined -- we only need to set those if defined.

**Pull request merged**